run-check.sh -> Haskell

RyanGlScott opened the pull request:
My attempt at resolving #32. I had to move some files to the `check/` directory in order to make this work smoothly with `cabal`, so please review to make sure that I didn't do anything bad.

sol commented on .gitignore:
```diff
@@ -1 +1,15 @@
+cabal.sandbox.config
+*.prof
+*.aux
+*.hp
```
I'm not excited about catch all ignore files.  If you want this behavior you could still do it in [your global git ignore file](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore).

sol commented on check/.gitignore:
```diff
@@ -0,0 +1 @@
+typediff
```
Will you ever have this file on your local machine?  I just `cabal install` `typediff` once and then have it in my path.

sol commented on check/run-check.sh:
```diff
@@ -0,0 +1,7 @@
+    wget https://github.com/haskell-compat/base-compat/releases/download/typediff-0.1.0/typediff
+    chmod +x typediff
+fi
+PATH=./:$PATH cabal test
```
Ok, ic, here we download `typediff`.  Maybe ditch that script altogether and put it in the readme that a user has to

    (cd ../typediff/ && cabal install)

before running the tests?  At least I don't really like to run scripts that download binary executables and run them without any form of trust metric.

sol commented on check/type-check.cabal:
```diff
@@ -0,0 +1,52 @@
+      base >= 4.3 && < 5
+    , process
+
+test-suite type-check-test
```
Would it make sense to make this a test suite on `base-compat` instead of putting it into a separate cabal package?

sol replied:
It would have the disadvantage that it e.g. doesn't build on Hackage, as it's not fully self contained due to `typediff`.  So maybe a separate cabal file is a better option.

sol commented on check/TypeCheck.hs:
```diff
@@ -0,0 +1,50 @@
+          , "Control.Monad.Compat"
+          , "Data.Foldable.Compat"
+          , "Data.List.Compat"
+          ]
```
What about inferring this list from the files we have in `check/index/`?

sol commented on check/TypeCheck.hs:
```diff
@@ -0,0 +1,50 @@
+                      output <- hGetContents hTypediff
+                      unless (null output) $ assertFailure output
+                  )
+                  mbStdoutH
```
Could all of this be simplified by using `readProcess`?  It requires a recent version of `process`, but I think it's cabal installable.

sol commented:
Except for my comments, if it works, looks good to me.

The rest of my comments are purely personal taste.  If you don't agree, feel free to ignore them.

sol commented on check/Setup.hs:
```diff
@@ -0,0 +1,2 @@
+import Distribution.Simple
+main = defaultMain
```
I like `Setup.lhs` as they can have a shebang and at the same time be a valid Haskell module.  I would even consider to symlink to `../Setup.lhs` (git can handle symlinks.)

sol commented on check/TypeCheck.hs:
```diff
@@ -0,0 +1,50 @@
+typeCheck :: FilePath -> String -> Spec
+typeCheck pwd module_ = describe module_
+                      . it "should have the expected type signatures"
+                      . withTempFile pwd "Temp.types" $ \fp h -> do
```
^ Personally I don't align in that way.  Firstly I only align by multiples of 2 (that would be for 4 for you I guess).  Secondly, you have to reallign everything if you e.g. add an other argument to `typeCheck` later on (or rename one, etc.).  My main issue with realligning is that you get bigger diffs.  This is harder to review and may lead to more conflicts.

For That reason, I would write the above code e.g. like so:

```haskell
typeCheck :: FilePath -> String -> Spec
typeCheck pwd module_ =
    describe module_
  . it "should have the expected type signatures"
  . withTempFile pwd "Temp.types" $ \fp h -> do
```

sol commented on check/Util.hs:
```diff
@@ -1,6 +1,6 @@
 module Util where
 
-import           System.Process
+import System.Process
```
I think aligning import statements like this is pretty common by now (and yes, contrary to what I said above, it's not a multiple of two).  I don't really care, I just adopted it at some point as others did it.

RyanGlScott commented:
Thanks for the feedback!

I didn't know about `readProcess` and `readProcessWithExitCode`, and luckily, they're available on pretty much every version of `process` that we'd be using.

I removed `run-check.sh` and changed the `README` to include the instructions on how to run the tests.

I think I incorporated all of your other stylistic suggestions, but let me know if I missed something.

sol commented:
LGTM
